/*
 * MIT License
 *
 * Copyright (c) 2020 - present Alexey Lapin
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package com.github.al.realworld.application.command

import com.github.al.realworld.api.command.CreateArticle
import com.github.al.realworld.api.command.CreateArticleResult
import com.github.al.realworld.application.ArticleAssembler
import com.github.al.realworld.application.service.SlugService
import com.github.al.realworld.bus.CommandHandler
import com.github.al.realworld.domain.model.Article
import com.github.al.realworld.domain.model.Tag
import com.github.al.realworld.domain.repository.ArticleRepository
import com.github.al.realworld.domain.repository.TagRepository
import com.github.al.realworld.domain.repository.UserRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.ZonedDateTime
import java.util.UUID

@Service
class CreateArticleHandler(
    private val articleRepository: ArticleRepository,
    private val tagRepository: TagRepository,
    private val userRepository: UserRepository,
    private val slugService: SlugService
) : CommandHandler<CreateArticleResult, CreateArticle> {

    @Transactional
    override fun handle(command: CreateArticle): CreateArticleResult {
        if (articleRepository.findByTitle(command.title).isPresent) {
            throw IllegalArgumentException("article [title=${command.title}] already exists")
        }

        val currentUser = userRepository.findByUsername(command.currentUsername)
            .orElseThrow { IllegalArgumentException("user [name=${command.currentUsername}] does not exist") }

        val now = ZonedDateTime.now()

        val article = Article(
            id = UUID.randomUUID(),
            slug = slugService.makeSlug(command.title),
            title = command.title,
            description = command.description,
            body = command.body,
            createdAt = now,
            updatedAt = now,
            author = currentUser
        )

        command.tagList?.forEach { tagName ->
            val tag = tagRepository.findByName(tagName).orElseGet { Tag(tagName) }
            article.tag(tag)
        }

        val savedArticle = articleRepository.save(article)

        return CreateArticleResult(ArticleAssembler.assemble(savedArticle, currentUser))
    }
}
